### DISCUSSION TITLE
How can I detect whether I'm in vi-mode or not?

### DISCUSSION THREAD

**Discussion by g-gundam** [lem-project/lem#1532] [Q&A] [answered]
In code, how can I query Lem to see if it's in vi-mode or not?

**Comment by cxxxr** [comment, marked as answer]
You can use `(lem:current-global-mode)`

**Comment by g-gundam** [comment]
## Switch Between Emacs and Vi Modes Quickly

Here's something similar in the spirit to evil's C-z which toggles evil's emacs state.  I put it on Ctrl-Shift-Z in Lem, because Ctrl-z was already taken by a bunch of frame commands.  It's useful in situations where something doesn't work well in vi-mode, and you need to quickly switch to emacs-mode.

```lisp
(in-package :lem-user)
(import 'lem:define-command)

(define-command toggle-between-emacs-and-vi () ()
  (let ((mode (lem:current-global-mode)))
    (if (typep mode 'lem-vi-mode/core:vi-mode)
        (lem-core::emacs-mode)
        (lem-vi-mode/core:vi-mode))))

(define-key *global-keymap* "C-Z" 'toggle-between-emacs-and-vi)
```

**Comment by vindarel** [comment]
related: to change vi mode's state we can do

```lisp
(lem-vi-mode:change-state 'lem-vi-mode:insert)
```
